examples/10-at-a-time: fix possible skipped final transfers

jay opened the pull request:
Prior to this change if curl_multi_perform returned 0 running handles and then all remaining transfers were added, then the perform loop would end immediately without performing those transfers.

Reported-by: Mikhail Kuznetsov

Fixes https://github.com/curl/curl/issues/9953
Closes #xxxx

bagder commented:
Alternative fix:
~~~diff
diff --git a/docs/examples/10-at-a-time.c b/docs/examples/10-at-a-time.c
index 1739a9e78..40ec52cd0 100644
--- a/docs/examples/10-at-a-time.c
+++ b/docs/examples/10-at-a-time.c
@@ -93,37 +93,39 @@ static size_t write_cb(char *data, size_t n, size_t l, void *userp)
   (void)data;
   (void)userp;
   return n*l;
 }
 
-static void add_transfer(CURLM *cm, int i)
+static void add_transfer(CURLM *cm, int i, int *left)
 {
   CURL *eh = curl_easy_init();
   curl_easy_setopt(eh, CURLOPT_WRITEFUNCTION, write_cb);
   curl_easy_setopt(eh, CURLOPT_URL, urls[i]);
   curl_easy_setopt(eh, CURLOPT_PRIVATE, urls[i]);
   curl_multi_add_handle(cm, eh);
+  (*left)++;
 }
 
 int main(void)
 {
   CURLM *cm;
   CURLMsg *msg;
   unsigned int transfers = 0;
   int msgs_left = -1;
-  int still_alive = 1;
+  int left = 0;
 
   curl_global_init(CURL_GLOBAL_ALL);
   cm = curl_multi_init();
 
   /* Limit the amount of simultaneous connections curl should allow: */
   curl_multi_setopt(cm, CURLMOPT_MAXCONNECTS, (long)MAX_PARALLEL);
 
   for(transfers = 0; transfers < MAX_PARALLEL; transfers++)
-    add_transfer(cm, transfers);
+    add_transfer(cm, transfers, &left);
 
   do {
+    int still_alive = 1;
     curl_multi_perform(cm, &still_alive);
 
     while((msg = curl_multi_info_read(cm, &msgs_left))) {
       if(msg->msg == CURLMSG_DONE) {
         char *url;
@@ -131,21 +133,22 @@ int main(void)
         curl_easy_getinfo(msg->easy_handle, CURLINFO_PRIVATE, &url);
         fprintf(stderr, "R: %d - %s <%s>\n",
                 msg->data.result, curl_easy_strerror(msg->data.result), url);
         curl_multi_remove_handle(cm, e);
         curl_easy_cleanup(e);
+        left--;
       }
       else {
         fprintf(stderr, "E: CURLMsg (%d)\n", msg->msg);
       }
       if(transfers < NUM_URLS)
-        add_transfer(cm, transfers++);
+        add_transfer(cm, transfers++, &left);
     }
-    if(still_alive)
+    if(left)
       curl_multi_wait(cm, NULL, 0, 1000, NULL);
 
-  } while(still_alive || (transfers < NUM_URLS));
+  } while(left);
 
   curl_multi_cleanup(cm);
   curl_global_cleanup();
 
   return EXIT_SUCCESS;

kuznetsov-m commented:
Yes, it works! Alternative fix also looks great!

jay commented:
> Alternative fix

Sure that works as well and fixes it in half the lines that I had changed. I think that the initial loop to add transfers should still be conditional on `transfers < MAX_PARALLEL && transfers < NUM_URLS` so I've added that.

> Yes, it works! Alternative fix also looks great!

Thanks for the report and testing.
